source/MRViewer/MRSpaceMouseHandlerHidapi.cpp

```diff
@@ -0,0 +1,201 @@
+#ifndef _WIN32
+#ifndef __EMSCRIPTEN__
+#include "MRSpaceMouseHandlerHidapi.h"
+
+#include "MRViewer/MRViewerFwd.h"
+#include "MRViewer.h"
+#include "MRGladGlfw.h"
+
+namespace MR
+{
+SpaceMouseHandlerHidapi::SpaceMouseHandlerHidapi()
+        : device_(nullptr)
+        , terminateListenerThread_(false)
+        , dataPacket_({0})
+        , packetLength_(0)
+{}
+
+SpaceMouseHandlerHidapi::~SpaceMouseHandlerHidapi()
+{
+    terminateListenerThread_ = true;
+    cv_.notify_one();
+
+    if ( listenerThread_.joinable() )
+        listenerThread_.join();
+
+    if ( device_ != nullptr )
+        hid_close( device_ );
+
+    hid_exit();
+}
+
+void SpaceMouseHandlerHidapi::initialize()
+{
+    if ( hid_init() ) {
+        spdlog::error( "HID API: init error" );
+        return;
+    }
+#ifdef __APPLE__
+    hid_darwin_set_open_exclusive( 0 );
+#endif
+
+#ifndef NDEBUG
+    hid_device_info* devs_ = hid_enumerate( 0x0, 0x0 );
+    printDevices_( devs_ );
+#endif
+    terminateListenerThread_ = false;
+    initListenerThread_();
+}
+
+bool SpaceMouseHandlerHidapi::findAndAttachDevice_() {
+    bool isDeviceFound = false;
+    for ( const auto& [vendorId, supportedDevicesId] : vendor2device_ )
+    {
+        // search through supported vendors
+        hid_device_info* localDevicesIt = hid_enumerate( vendorId, 0x0 );
+        while( localDevicesIt && !isDeviceFound )
+        {
+            for ( ProductId deviceId: supportedDevicesId )
+            {
+                if (  deviceId == localDevicesIt->product_id )
+                {
+                    device_ = hid_open( vendorId, deviceId, NULL );
+                    if ( device_ )
+                    {
+                        isDeviceFound = true;
+                        spdlog::info( "SpaceMouse Found: type: {} {} path: {} ", vendorId, deviceId, localDevicesIt->path );
+                        break;
+                    }
+                    else
+                    {
+                        spdlog::error( "HID API: device open error" );
+                    }
+                }
+            }
+            localDevicesIt = localDevicesIt->next;
+        }
+        hid_free_enumeration( localDevicesIt );
+    }
+    return isDeviceFound;
+}
+
+void SpaceMouseHandlerHidapi::handle()
+{
+    // works in pair with SpaceMouseHandlerHidapi::startListenerThread_()
+    std::unique_lock<std::mutex> syncThreadLock( syncThreadMutex_, std::defer_lock );
+    if ( !syncThreadLock.try_lock() )
+        return;
+
+    if ( packetLength_ <= 0 )
+        return;
+
+    auto &viewer = getViewerInstance();
+    // process saved data packet
+    Vector3f translate, rotate;
+    convertInput_( dataPacket_, packetLength_, translate, rotate );
+    viewer.spaceMouseMove( translate, rotate );
```

Comment:
do we need to check `packetLen == 13 || packetLen == 7` here? 
mb make functoin not to duplicate
```c++
    Vector3f translate, rotate;
    convertInput_( dataPacket_, packetLength_, translate, rotate );
    viewer.spaceMouseMove( translate, rotate );

Reply:
or hide it inside `convertInput_`